Upload binaries during CI build (#374)
open-telemetry/opentelemetry-dotnet-instrumentation: docs/troubleshooting.md

# Troubleshooting

> Check if you are not hitting one of the issues listed below.

Reviewer: ```suggestion
Make sure that you are not hitting one of following issues:
```

> ## Handling of Assembly version Conflicts

Reviewer: ```suggestion
## Handling of assembly version conflicts
```

OpenTelemetry SDK NuGet package are deployed with the OpenTelemetry .NET Instrumentation.

> Conflicts in assemblies referenced by "source instrumentations", should be handled in
> the same way: updating the project references to ensure that they are the same version
> as the one used by the instrumentation.

Reviewer: ```suggestion
To handle conflicts in assemblies referenced by "source instrumentations", update the project 
references to ensure that they are on the same versions as the ones used by the instrumentation.
```

> However, the workarounds proposed above only work at build time. When a rebuild is not
> possible, there are ways to force the application to use the assembly versions shipped
> together with the instrumentation.

Reviewer: ```suggestion
Previous workarounds only work at build time. When a rebuild is not possible, use one of the
following suggestions to force the application to use the assembly versions shipped with the 
instrumentation.
```

> For .NET Framework applications, the workaround is to use Binding Redirects. For .NET Core

Reviewer: ```suggestion
For .NET Framework applications, the workaround is to use binding redirects. For .NET Core
```

[Framework-dependent deployment](https://docs.microsoft.com/en-us/dotnet/core/deploying/deploy-with-vs?tabs=vs156#framework-dependent-deployment)
applications,

> [Additional-deps](https://github.com/dotnet/runtime/blob/main/docs/design/features/additional-deps.md)

Reviewer: ```suggestion
[Additional-deps](https://github.com/dotnet/runtime/blob/main/docs/design/features/additional-deps.md),
```

and [runtime package store](https://docs.microsoft.com/en-us/dotnet/core/deploying/runtime-store)

> from OpenTelemetry .NET Auto-Instrumentation installation location can be used as workaround.

Reviewer: ```suggestion
from OpenTelemetry .NET, use the auto-instrumentation installation path.
```

> For other .NET Core deployment models, the workaround is to manipulate the `deps.json`. Currently,
> we do not automate any of these workarounds, but we are manually validating them.

Reviewer: ```suggestion
For other .NET Core deployment models, edit the `deps.json` file.
```

I wouldn't mention the lack of automation in Troubleshooting, but move it to a development or contributing guide (or even a TO-DO).

> ### .NET Framework Binding Redirects

Reviewer: Not sure about the capitalization here. I'd use "binding redirects". Same for "dependency file".

Author: Addressed https://github.com/open-telemetry/opentelemetry-dotnet-instrumentation/pull/374/commits/913242cbf46ebd77c2a7b96e0dbc6ef173768c9f

The [samples/BindingRedirect](./../samples/BindingRedirect/) app shows how

> to use the `app.config` to solve the version conflicts. As configured in the PoC branch,

Reviewer: ```suggestion
to use the `app.config` file to solve version conflicts. As configured in the PoC branch,
```

> the BindingRedirect sample can only run successfully under the instrumentation since the

Author: ```suggestion
The sample can only run successfully under the instrumentation since the
```

> the BindingRedirect sample can only run successfully under the instrumentation since the
> binding redirect makes the application dependent on a version of `System.Diagnostics.DiagnosticSource`
> that is not available during building time.

Reviewer: ```suggestion
the BindingRedirect sample can only run successfully under the instrumentation, as the
binding redirect makes the application dependent on a version of `System.Diagnostics.DiagnosticSource`
that is not available at build time.
```

### .NET Core Dependency File

> To fix assembly version conflicts in .NET Core, the `<application>.deps.json`, generated by default
> during the build of .NET Core applications, needs to be modified. Build a .NET Core app with package
> references using the required version and use the respective `deps.json` file to see what changes
> are needed.

Reviewer: ```suggestion
To fix assembly version conflicts in .NET Core, edit the default `<application>.deps.json` file 
generated at build for .NET Core applications. Build a .NET Core app with package
references using the required version, and use the respective `deps.json` file to see what changes
are needed.
```

> To experiment with modifying the `deps.json` file, add a reference to the required version OpenTelemetry
> package to the [samples/CoreAppOldReference](./../samples/CoreAppOldReference/) sample and rebuild the
> application. Save the generated `deps.json` file, remove the package reference and rebuild the
> sample app. Compare the files to understand the changes.

Reviewer: ```suggestion
To test your edits to  the`deps.json` file, add a reference to the required version of the OpenTelemetry
package to the [samples/CoreAppOldReference](./../samples/CoreAppOldReference/) sample and rebuild the
application. Save the generated `deps.json` file, remove the package reference, and rebuild the
sample app. Compare the files to explore the changes.
```

## No proper relationship between spans

On .NET Framework strong name signing can force multiple versions
of the same assembly being loaded on the same process.
This causes a separate hierarchy of Activity objects.
If you are referencing packages in your application that use
different version of the `System.Diagnostics.DiagnosticSource`
than the `OpenTelemetry.Api` used by
OpenTelemetry .NET Auto-Instrumentation (`6.0.0`)
you have to explicitly reference
the `System.Diagnostics.DiagnosticSource` package in the correct version
in your application.
This will cause automatic binding redirection to occur resolving the issue.

If automatic binding redirection is [disabled](https://docs.microsoft.com/en-us/dotnet/framework/configure-apps/how-to-enable-and-disable-automatic-binding-redirection)
you can also manually add binding redirection to [the `App.config` file](../samples/BindingRedirect/App.config).

## High CPU usage

Check if you have not enabled the auto-instrumentation globally

> by setting theenvironment variables on system or user level.

Reviewer 2: ```suggestion
by setting the environment variables on system or user level.
```

If it was really intended, `OTEL_DOTNET_AUTO_EXCLUDE_PROCESSES` and `OTEL_DOTNET_AUTO_INCLUDE_PROCESSES`
to include/exclude applications from the tracing auto-instrumentation.

> ## Investigating other issues

Reviewer: ```suggestion
## Collect debug logs
```
